_posts/2018-05-10-Rust-1.26.md

```diff
@@ -0,0 +1,545 @@
+in that logic.
+
+In Rust 1.26, you can now declare `main` that returns `Result`:
+
+```rust
+use std::fs::File;
+
+fn main() -> Result<(), std::io::Error> {
+    let f = File::open("bar.txt")?;
+
+    Ok(())
+}
+```
+
+This now works just fine! If `main` returns an error, this will
+exit with an error code, and print out a debug representation
+of the error.
+
+#### Inclusive ranges with `..=`
+
+Since well before Rust 1.0, you've been able to create exclusive ranges with `..`
+like this:
+
+```rust
+for i in 1..3 {
+    println!("i: {}", i);
+}
+```
+
+This will print `i: 1` and then `i: 2`. In Rust 1.26, you can now create an
+inclusive range, like this:
+
+```rust
+for i in 1..=3 {
+    println!("i: {}", i);
+}
+```
+
+This will print `i: 1` and then `i: 2` like before, but also `i: 3`; the
+three is included in the range. Inclusive ranges are especially useful
+if you want to iterate over every possible value in a range. For example,
+this is a surprising Rust program:
+
+```rust
+fn takes_u8(x: u8) {
+    // ...
+}
+
+fn main() {
+    for i in 0..256 {
+        println!("i: {}", i);
+        takes_u8(i);
+    }
+}
+```
+
+What does this program do? The answer: nothing. The warning we get when
+compiling has a hint:
+
+```text
+warning: literal out of range for u8
+ --> src/main.rs:6:17
+  |
+6 |     for i in 0..256 {
+  |                 ^^^
+  |
+  = note: #[warn(overflowing_literals)] on by default
+```
+
+That's right, since `i` is a `u8`, this overflows, and is the same as writing
+`for i in 0..0`, so the loop executes zero times.
+
+We can do this with inclusive ranges, however:
+
+```rust
+fn takes_u8(x: u8) {
+    // ...
+}
+
+fn main() {
+    for i in 0..=255 {
+        println!("i: {}", i);
+        takes_u8(i);
+    }
+}
+```
+
+This will produce those 256 lines of output you might have been expecting.
+
+#### Basic slice patterns
+
+Another long-awaited feature is "slice patterns." These let you match on
+slices similar to how you match on other data types. For example:
+
+```rust
+let arr = [1, 2, 3];
+
+match arr {
+    [1, _, _] => "starts with one",
+    [a, b, c] => "starts with something else",
+}
+```
+
+In this case, we know `arr` has a length of three, and so we need three entries
+inside the `[]`s. We can also match when we don't know the length:
+
+```rust
+fn foo(s: &[u8]) {
+    match s {
+        [a, b] => (),
+        [a, b, c] => (),
+        _ => (),
+    }
+}
+```
+
+Here, we don't know how long `s` is, so we can write the first two arms, each with
+different lengths. This also means we need a `_` term, since we aren't covering
+every possible length, nor could we!
+
+#### Speed improvements
+
+We continue to work on the speed of the compiler. We discovered that deeply
+nesting types was non-linear in some cases, and [a fix was
+implemented](https://github.com/rust-lang/rust/pull/48296). We're seeing up
+to a 12% reduction in compile times from this change, but many other smaller
```

Comment:
Technically, our benchmarks on perf have a maximum decrease of 17.6%... but some cases could have even larger decreases, in theory.